File: src/games/sudoku/gameplay/analysis/techniqueModules/pointingPairTriple.ts
import { sudokuDigits } from '../../../types';
import { boxCellIndexes, cellBoxIndexes, cellColIndexes, cellRowIndexes } from '../bitmask';
import { buildCandidateEliminationMove, getHouseDigitMatches } from '../techniqueHelpers';
import type { SudokuTechniqueDispatcher } from '../techniqueModuleTypes';
import type { SudokuCanonicalMove } from '../moves';

export const pointingPairTripleTechnique: SudokuTechniqueDispatcher = {
  technique: 'pointing-pair-triple',
  tier: 'medium',
  findMove(state) {
    let best: SudokuCanonicalMove | null = null;
    let bestComplexity = Infinity;

    for (let boxIndex = 0; boxIndex < boxCellIndexes.length; boxIndex += 1) {
      const boxCells = boxCellIndexes[boxIndex];
      const complexity = boxCells.filter((i) => state.board[i] === 0).length;
      if (complexity >= bestComplexity) {
        continue;
      }

      for (const digit of sudokuDigits) {
        const matches = getHouseDigitMatches(state, boxCells, digit);
        if (matches.length < 2 || matches.length > 3) {
          continue;
        }

        const distinctRows = Array.from(new Set(matches.map((index) => cellRowIndexes[index])));
        if (distinctRows.length === 1) {
          const targetRow = distinctRows[0];
          const move = buildCandidateEliminationMove({
            technique: 'pointing-pair-triple',
            eliminations: Array.from({ length: 9 }, (_, col) => (targetRow * 9) + col)
              .filter((index) => cellBoxIndexes[index] !== boxIndex)
              .filter((index) => state.board[index] === 0 && (state.candidateMask[index] & (1 << (digit - 1))) !== 0)
              .map((index) => ({ index, digit })),
            evidenceCells: matches,
            houses: [
              { kind: 'box', index: boxIndex },
              { kind: 'row', index: targetRow },
            ],
            complexity,
          });
          if (move) {
            best = move;
            bestComplexity = complexity;
            break;
          }
        }

        const distinctCols = Array.from(new Set(matches.map((index) => cellColIndexes[index])));
        if (distinctCols.length === 1) {
          const targetCol = distinctCols[0];
          const move = buildCandidateEliminationMove({
            technique: 'pointing-pair-triple',
            eliminations: Array.from({ length: 9 }, (_, row) => (row * 9) + targetCol)
              .filter((index) => cellBoxIndexes[index] !== boxIndex)
              .filter((index) => state.board[index] === 0 && (state.candidateMask[index] & (1 << (digit - 1))) !== 0)
              .map((index) => ({ index, digit })),
            evidenceCells: matches,
            houses: [
              { kind: 'box', index: boxIndex },
              { kind: 'column', index: targetCol },
            ],
            complexity,
          });
          if (move) {
            best = move;
            bestComplexity = complexity;
            break;
          }
        }
      }
    }

    return best;
  },
};

File: engine/src/main/kotlin/com/quietgrid/engine/sudoku/techniques/PointingPairTriple.kt
package com.quietgrid.engine.sudoku.techniques

import com.quietgrid.engine.core.Difficulty
import com.quietgrid.engine.sudoku.*

object PointingPairTripleTechnique : SudokuTechniqueDispatcher {
    override val technique = SudokuTechnique.POINTING_PAIR_TRIPLE
    override val tier = Difficulty.MEDIUM

    override fun findMove(state: SudokuBitmaskState): SudokuCanonicalMove? {
        var best: SudokuCanonicalMove? = null
        var bestComplexity = Int.MAX_VALUE

        for (boxIndex in boxCellIndexes.indices) {
            val boxCells = boxCellIndexes[boxIndex]
            val complexity = boxCells.count { state.board[it] == 0 }
            if (complexity >= bestComplexity) continue

            for (digit in 1..9) {
                val matches = getHouseDigitMatches(state, boxCells, digit)
                if (matches.size !in 2..3) continue

                val distinctRows = matches.map { cellRowIndexes[it] }.distinct()
                if (distinctRows.size == 1) {
                    val targetRow = distinctRows[0]
                    val eliminations = rowCellIndexes[targetRow]
                        .filter { cellBoxIndexes[it] != boxIndex && state.board[it] == 0 && (state.candidateMask[it] and digitToBit[digit]) != 0 }
                        .map { it to digit }
                    val move = buildCandidateEliminationMove(
                        SudokuTechnique.POINTING_PAIR_TRIPLE, eliminations, matches,
                        listOf(SudokuHouseRef("box", boxIndex), SudokuHouseRef("row", targetRow)), complexity,
                    )
                    if (move != null) { best = move; bestComplexity = complexity; break }
                }

                val distinctCols = matches.map { cellColIndexes[it] }.distinct()
                if (distinctCols.size == 1) {
                    val targetCol = distinctCols[0]
                    val eliminations = columnCellIndexes[targetCol]
                        .filter { cellBoxIndexes[it] != boxIndex && state.board[it] == 0 && (state.candidateMask[it] and digitToBit[digit]) != 0 }
                        .map { it to digit }
                    val move = buildCandidateEliminationMove(
                        SudokuTechnique.POINTING_PAIR_TRIPLE, eliminations, matches,
                        listOf(SudokuHouseRef("box", boxIndex), SudokuHouseRef("column", targetCol)), complexity,
                    )
                    if (move != null) { best = move; bestComplexity = complexity; break }
                }
            }
        }
        return best
    }
}
